### PR TITLE
Pr merlin algorithm

### LINKED ISSUE #519: TTP: MerlinAlgorithm

### PR DESCRIPTION
**Pull request by emailweixu**
Also fixed a few bugs:
1. out-of-memory during play for RNN model. Need to run under torch.no_grad() context
2. checkpoint loading when batch normalization layer is used. It has its own _load_from_state_dict()
3. some bugs in suite_dmlab_test

### CODE REVIEW

File: alf/algorithms/algorithm.py [hunk 1]
```diff
@@ -432,11 +432,24 @@ def load(module, prefix='', visited=None):
...
+                module._load_from_state_dict(
+                    state_dict, prefix, local_metadata, True, missing_keys,
+                    unexpected_keys, error_msgs, visited)
+            else:
+                # Some pytorch modules (e.g. BatchNorm layers) override
+                # _load_from_state_dict, which uses the original
+                # Module._load_from_state_dict. So we have to handle them
+                # differently. Not using `visited` should not cause a problem
+                # because those modules are not implemented by ALF and will have
+                # have cycle through them.
```
Comment (Haichao-Zhang) [inline alf/algorithms/algorithm.py:447]: ```have have cycle``` -> do you mean```not have cycle```?

File: alf/trainers/policy_trainer.py [hunk 2]
```diff
@@ -319,13 +319,14 @@ def play(root_dir,
...
-        time_step, policy_state = _step(
-            algorithm=algorithm,
-            env=env,
-            time_step=time_step,
-            policy_state=policy_state,
-            epsilon_greedy=epsilon_greedy,
-            metrics=[])
+        with torch.no_grad():
+            time_step, policy_state = _step(
+                algorithm=algorithm,
```
Comment (Haichao-Zhang) [inline alf/trainers/policy_trainer.py:324]: Since batch-norm might be used in algorithm, I think here we may need to turn on the ```eval``` mode by using ```algorithm.eval()```, which will freeze the batch-norm statistics. 
We might can also do the similar thing for the periodic evaluation during training, but this might not be very critical and can be done in the future.

Comment (emailweixu) [reply alf/trainers/policy_trainer.py:324]: I am aware of this. We can do it in the future.

File: alf/algorithms/actor_critic_algorithm.py [hunk 3]
```diff
@@ -65,7 +65,7 @@ def __init__(self,
             loss (None|ActorCriticLoss): an object for calculating loss. If
                 None, a default loss of class loss_class will be used.
             loss_class (type): the class of the loss. The signature of its
-                constructor: loss_class(action_spec, debug_summaries)
```
Comment (hnyu) [inline alf/algorithms/actor_critic_algorithm.py:68]: docstring of this file needs update

File: alf/algorithms/decoding_algorithm.py [hunk 4]
```diff
@@ -13,28 +13,30 @@
...
                  loss_weight=1.0):
-        """Create a decoding algorithm.
+        """
 
         Args:
-            decoder (Network)
-            loss (Callable): loss function with signature loss(y_true, y_pred)
+            decoder (Network): network for decoding target from input.
+            loss (Callable): loss function with signature ``loss(y_pred, y_true)``.
+                Note that is should not reduce. It should at least keep the
```
Comment (hnyu) [inline alf/algorithms/decoding_algorithm.py]: is -> it

File: alf/algorithms/decoding_algorithm.py [hunk 5]
```diff
@@ -44,12 +46,12 @@ def __init__(self,
...
         """Train one step.
 
         Args:
-            inputs (tuple): tuple of (inputs, target)
-            state (nested Tensor): network state for `decoder`
+            inputs (tuple): tuple of (input, target)
+            state (nested Tensor): network state for ``decoder``
 
         Returns:
             AlgorithmStep with the following fields:
```
Comment (hnyu) [inline alf/algorithms/decoding_algorithm.py]: Return format needs correction

File: alf/algorithms/memory.py [hunk 6]
```diff
@@ -155,20 +155,39 @@ def build(self, batch_size):
...
+    def create_keynet(self, query_spec, num_keys):
+        """Create a net which can be used to generate keys.
+
+        The created keynet can be used with ``genkey_and_read``.
+
+        Args:
+            query_spec (alf.TensorSpec): the spec for the query
+            num_keys (int): the number of keys to be generated.
+        Returns:
+            a function which calculates ``num_keys`` keys given query.
```
Comment (hnyu) [inline alf/algorithms/memory.py]: Callable: a function which ...

File: alf/algorithms/memory.py [hunk 7]
```diff
@@ -155,20 +155,39 @@ def build(self, batch_size):
...
+                (batch_size, num_keys * (dim + 1)). ``keynet`` can be created
+                using ``create_keynet``.
             query (Tensor): the query from which the keys are generated
             flatten_result (bool): If True, the result shape will be
-               (batch_size, num_keys * dim), otherwise it is
-               (batch_size, num_keys, dim)
+                (batch_size, num_keys * dim), otherwise it is
+                (batch_size, num_keys, dim)
         Returns:
             resutl Tensor: If flatten_result is True,
```
Comment (hnyu) [inline alf/algorithms/memory.py]: resutl Tensor -> Tensor

File: alf/algorithms/merlin_algorithm.py [hunk 8]
```diff
@@ -254,92 +242,97 @@ def __init__(self,
...
+        action_distribution, _ = self._actor_net(policy_input)
+        return action_distribution, rnn_state
+
+    def rollout_step(self, time_step: TimeStep, state):
         """Train one step.
 
         Args:
-            time_step: time_step.observation should be the latent vector
-            state: state of the model
+            time_step (TimeStep): time_step.observation should be the latent
```
Comment (hnyu) [inline alf/algorithms/merlin_algorithm.py]: \``time_step.observation\``

File: alf/algorithms/merlin_algorithm.py [hunk 9]
```diff
@@ -384,13 +382,22 @@ def __init__(self,
             action_spec (nested BoundedTensorSpec): representing the actions.
             encoders (nested Network): the nest should match observation_spec
             decoders (nested Algorithm): the nest should match observation_spec
+            env (Environment): The environment to interact with. `env` is a
```
Comment (hnyu) [inline alf/algorithms/merlin_algorithm.py]: single quote to double quotes

File: alf/algorithms/vae.py [hunk 10]
```diff
@@ -50,84 +49,78 @@ def __init__(self,
                  z_dim,
                  input_tensor_spec: TensorSpec = None,
                  preprocess_network: EncodingNetwork = None,
-                 z_mean_prior_network: EncodingNetwork = None,
-                 z_log_var_prior_network: EncodingNetwork = None,
+                 z_prior_network: EncodingNetwork = None,
                  beta=1.0,
                  name="VariationalAutoEncoder"):
         """Create an instance of `VariationalAutoEncoder`.
```
Comment (hnyu) [inline alf/algorithms/vae.py]: remove "create ..."

File: alf/algorithms/vae.py [hunk 11]
```diff
@@ -50,84 +49,78 @@ def __init__(self,
...
 
         Args:
             inputs (nested Tensor): if a prior network is provided, this is a
-                tuple of `(prior_input, new_observation)`.
+                tuple of ``(prior_input, new_observation)``.
 
         Returns:
-            (z, kl_loss): `z` is a tensor of shape (`N`, `z_dim`), `kl_loss` is
-                a tensor of shape (`N`,).
+            (z, kl_loss): ``z`` is a tensor of shape (``B``, ``z_dim``), ``kl_loss``
```
Comment (hnyu) [inline alf/algorithms/vae.py]: tuple:
- z (Tensor): ...
- kl_loss (Tensor): ...

File: alf/algorithms/vae.py [hunk 12]
```diff
@@ -50,84 +49,78 @@ def __init__(self,
...
+                is a tensor of shape (``B``,).
         """
-        if self._z_mean_prior_network:
+        if self._z_prior_network:
             prior_input, new_obs = inputs
-            prior_z_mean, _ = self._z_mean_prior_network(prior_input)
-            prior_z_log_var, _ = self._z_log_var_prior_network(prior_input)
-            inputs = [prior_input, new_obs, prior_z_mean, prior_z_log_var]
+            prior_z_mean_and_log_var, _ = self._z_prior_network(prior_input)
```
Comment (hnyu) [inline alf/algorithms/vae.py:115]: Is it OK to assume that z_mean and z_log_var share parameters up to the second last layer?

File: alf/algorithms/vae.py [hunk 13]
```diff
@@ -140,24 +133,25 @@ def _sampling_forward(self, inputs):
...
         Args:
             inputs (nested Tensor): data to be encoded. If there is a prior
-                network, then `inputs` is a tuple of `(prior_input, new_obs)`.
+                network, then ``inputs`` is a tuple of ``(prior_input, new_obs)``.
             state (Tensor): empty tuple ()
 
         Returns:
             AlgStep:
-                output: the latent vector `z`
+                output: the latent vector ``z``
```
Comment (hnyu) [inline alf/algorithms/vae.py]: should add '-' to each field, otherwise the lines are concatenated

File: alf/algorithms/vae.py [hunk 14]
```diff
@@ -18,28 +18,27 @@
...
     between posterior and prior.
 
     Mathematically:
 
-    log p(x) >= E_z log P(x|z) - beta KL(q(z|x) || prior(z))
+    :math:`\log p(x) >= E_z \log P(x|z) - \beta KL(q(z|x) || prior(z))`
 
-    `sampling_forward()` method returns sampled z and KL, it is up to user of
-    this class to use returned z to decode and compute reconstructive loss to
+    ``forward()`` method returns sampled z and KLD, it is up to the user of
```
Comment (hnyu) [inline alf/algorithms/vae.py]: sampling_forward?

File: alf/environments/suite_dmlab.py [hunk 15]
```diff
@@ -52,14 +52,14 @@ def action_discretize(action_spec,
...
     original deepmind lab environment action_spec:
-
-    `[{'max': 512, 'min': -512, 'name': 'LOOK_LEFT_RIGHT_PIXELS_PER_FRAME'},
-    {'max': 512, 'min': -512, 'name': 'LOOK_DOWN_UP_PIXELS_PER_FRAME'},
-    {'max': 1, 'min': -1, 'name': 'STRAFE_LEFT_RIGHT'},
-    {'max': 1, 'min': -1, 'name': 'MOVE_BACK_FORWARD'},
-    {'max': 1, 'min': 0, 'name': 'FIRE'},
-    {'max': 1, 'min': 0, 'name': 'JUMP'},
-    {'max': 1, 'min': 0, 'name': 'CROUCH'}]`
+    .. code-block:: python
```
Comment (hnyu) [inline alf/environments/suite_dmlab.py:56]: Leave an empty line here

File: alf/networks/rnn.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            self._cells.append(nn.LSTMCell(input_size, hidden_size))
+            spec = alf.TensorSpec(shape=(hidden_size, ))
+            self._state_spec.append((spec, spec))
+            input_size = hidden_size
+
+    @property
+    def state_spec(self):
+        return self._state_spec
+
+    def forward(self, inputs, state):
```
Comment (hnyu) [inline alf/networks/rnn.py]: comment

File: alf/networks/rnn.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+import gin
+import torch
+import torch.nn as nn
+
+import alf
+from .network import Network
+
+
+@gin.configurable(whitelist=['output_layers'])
+class StackedLSTMCell(Network):
```
Comment (hnyu) [inline alf/networks/rnn.py]: Can this be integrated into LSTMEncodingNetwork?

**Review by hnyu — CHANGES_REQUESTED** [review]
mostly format

File: alf/trainers/policy_trainer.py (hunk 2 above)
Comment (emailweixu) [reply alf/trainers/policy_trainer.py:324]: And it seems that pytorch handles batch_size=1 correctly, since it can correctly play even without using eval()

Comment (emailweixu) [reply alf/trainers/policy_trainer.py:324]: Need to check the source code to confirm though

File: alf/networks/rnn.py (hunk 17 above)
Comment (emailweixu) [reply alf/networks/rnn.py]: integrated.

File: alf/environments/suite_dmlab.py (hunk 15 above)
Comment (emailweixu) [reply alf/environments/suite_dmlab.py:56]: done

File: alf/algorithms/vae.py (hunk 14 above)
Comment (emailweixu) [reply alf/algorithms/vae.py]: changed to train_step

File: alf/algorithms/vae.py (hunk 13 above)
Comment (emailweixu) [reply alf/algorithms/vae.py]: done

File: alf/algorithms/memory.py (hunk 7 above)
Comment (emailweixu) [reply alf/algorithms/memory.py]: done

File: alf/algorithms/decoding_algorithm.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/decoding_algorithm.py]: updated

File: alf/algorithms/memory.py (hunk 6 above)
Comment (emailweixu) [reply alf/algorithms/memory.py]: done

File: alf/algorithms/decoding_algorithm.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/decoding_algorithm.py]: done

File: alf/trainers/policy_trainer.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/trainers/policy_trainer.py:324]: I think during play, the initial running mean and std of batch-norm are constructed by loading from checkpoint, therefore it can run without error for batch_size =1. But without eval, the statistics are still changing.
Also for training, I guess if we set ```num_parallel_environments=1``` there will be problem. We can do more investigation in the future in another PR.

**Review by Haichao-Zhang — DISMISSED** [review]

**Comment by hnyu** [comment]
Seems that there are several comments accidentally missed

File: alf/algorithms/merlin_algorithm.py (hunk 8 above)
Comment (emailweixu) [reply alf/algorithms/merlin_algorithm.py]: done

File: alf/algorithms/merlin_algorithm.py (hunk 9 above)
Comment (emailweixu) [reply alf/algorithms/merlin_algorithm.py]: done

File: alf/algorithms/vae.py (hunk 10 above)
Comment (emailweixu) [reply alf/algorithms/vae.py]: removed

File: alf/algorithms/vae.py (hunk 11 above)
Comment (emailweixu) [reply alf/algorithms/vae.py]: done

File: alf/algorithms/vae.py (hunk 12 above)
Comment (emailweixu) [reply alf/algorithms/vae.py:115]: This way makes the config simpler. If there is a need to separate branch for mean and log_var, users can always construct their own network.

**Review by hnyu — APPROVED** [review]
